language/vm/vm_runtime/src/code_cache/module_cache.rs

```diff
@@ -294,20 +294,28 @@ impl<'alloc> VMModuleCache<'alloc> {
                     fields,
                 } => {
                     let mut field_types = vec![];
-                    for field in module.field_def_range(*field_count, *fields) {
-                        let ty = try_runtime!(self.resolve_signature_token_with_fetcher(
-                            module,
-                            &module.type_signature_at(field.signature).0,
-                            gas_meter,
-                            fetcher
-                        ));
-                        if let Some(t) = ty {
-                            field_types.push(t);
-                        } else {
-                            return Ok(Ok(None));
+                    match module.field_def_range(*field_count, *fields) {
+                        Ok(field_defs) => {
+                            for field in field_defs {
+                                let ty = try_runtime!(self.resolve_signature_token_with_fetcher(
+                                    module,
+                                    &module.type_signature_at(field.signature).0,
+                                    gas_meter,
+                                    fetcher
+                                ));
+                                if let Some(t) = ty {
+                                    field_types.push(t);
+                                } else {
+                                    return Ok(Ok(None));
+                                }
+                            }
+                            StructDef::new(field_types)
                         }
+                        Err(VMStaticViolation::RangeOutOfBounds(_, _, _, _)) => {
+                            return Err(VMInvariantViolation::LinkerError)
```

Comment:
It seems a bit strange to map the RangeOutOfBounds error in this (lossy) way.

Reply:
Fixed